clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -734,11 +740,11 @@ private ClusterResourceListeners configureClusterResourceListeners(Serializer<K>
     private int partition(ProducerRecord<K, V> record, byte[] serializedKey, byte[] serializedValue, Cluster cluster) {
         Integer partition = record.partition();
         if (partition != null) {
-            List<PartitionInfo> partitions = cluster.partitionsForTopic(record.topic());
-            int lastPartition = partitions.size() - 1;
+            int lastPartition = cluster.partitionsForTopic(record.topic()).size() - 1;
             // they have given us a partition, use it
             if (partition < 0 || partition > lastPartition) {
-                throw new IllegalArgumentException(String.format("Invalid partition given with record: %d is not in the range [0...%d].", partition, lastPartition));
+                throw new UnknownTopicOrPartitionException(
```

Comment:
Using `UnknownTopicOrPartitionException` was an overlook, since it's a `RetriableException`. Perhaps `InvalidPartitionsException` would be more appropriate (I didn't like the plural), which is also an `ApiException`, but not retriable one. From what I can tell `ApiException`s are thrown in a number of situations that could also lead to the connector task failing to send the record, and without appropriate provision in the callback, not quitting. I believe such exceptions are also `InvalidTopicException`, `RecordTooLargeException` and maybe others. 

However, I agree that at this point using `ApiException` would change how producer reacts when partition is out of bounds. As a middle-ground solution, I defined a `PartitionOutOfBoundsException` that is a `KafkaException` and can be used to specifically handle situations where we know that such a partition will be available at some point, and can be used to wait or extend the partition range. 

Let me know if you think that's an overkill, and if throwing a `KafkaException` with an appropriate message is preferable.